homeassistant/components/homekit/aidmanager.py

```diff
@@ -0,0 +1,142 @@
+"""
+Manage allocation of accessory ID's.
+
+HomeKit needs to allocate unique numbers to each accessory. These need to
+be stable between reboots and upgrades.
+
+Using a hash function to generate them means collisions. It also means you
+can't change the hash without causing breakages for HA users.
+
+This module generates and stores them in a HA storage.
+"""
+import logging
+import random
+from zlib import adler32
+
+from fnvhash import fnv1a_32
+
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_registry import RegistryEntry
+from homeassistant.helpers.storage import Store
+
+from .const import DOMAIN
+
+AID_MANAGER_STORAGE_KEY = f"{DOMAIN}.aids"
+AID_MANAGER_STORAGE_VERSION = 1
+AID_MANAGER_SAVE_DELAY = 2
+
+INVALID_AIDS = (0, 1)
+
+AID_MIN = 2
+AID_MAX = 18446744073709551615
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def get_system_unique_id(entity: RegistryEntry):
+    """Determine the system wide unique_id for an entity."""
+    return f"{entity.platform}.{entity.domain}.{entity.unique_id}"
+
+
+def _generate_aids(unique_id: str, entity_id: str) -> int:
+    """Generate accessory aid."""
+
+    # Backward compatibility: Previously HA used to *only* do adler32 on the entity id.
+    # Not stable if entity ID changes
+    # Not robust against collisions
+    yield adler32(entity_id.encode("utf-8"))
+
+    # Use fnv1a_32 of the unique id as
+    # fnv1a_32 has less collisions than
+    # adler32
+    yield fnv1a_32(unique_id.encode("utf-8"))
+
+    # If called again resort to random allocations.
+    # Given the size of the range its unlikely we'll encounter duplicates
+    # But try a few times regardless
+    for _ in range(5):
+        yield random.randrange(AID_MIN, AID_MAX)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Standard pseudo-random generators are not suitable for security/cryptographic purposes.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5382946)